### PR TITLE
Implemented tools for Deep Linking

### PR DESCRIPTION
**Pull request by JosXa**
- Implemented new helper-method `create_deep_linked_url`
- Added `Filters.regex` and documented that it is not yet compliant with #835 (i.e. not passing groups or groupdict)
- Tests for all above
- Added example for deep linking
- Modified examples so that they all get their token from a `settings.py`. That way, you can set your token once and access all of the example bots. Added documentation to this in the appropriate places, especially that `Updater(TOKEN)`, where TOKEN stems from the `settings.py`, requires you to replace the constant with a **string** if you users want to use the example standalone. This might be controversial, so please respond if you think it is a good addition.

### CODE REVIEW

File: telegram/ext/filters.py [hunk 1]
```diff
@@ -171,6 +173,33 @@ def filter(self, message):
...
+        """
+
+        def __init__(self, pattern):
+            if isinstance(pattern, string_types):
+                pattern = re.compile(pattern)
+            self.pattern = pattern
+            self.name = 'Filters.regex({})'.format(self.pattern)
+
+        def filter(self, message):
+            return bool(re.search(self.pattern, message.text))
```
Comment (tsnoam) [inline telegram/ext/filters.py]: Is there a reason not to do:
 ```python
return bool(self.pattern.search(message.text))
```

File: telegram/ext/filters.py [hunk 2]
```diff
@@ -171,6 +173,33 @@ def filter(self, message):
...
+
+
```
Comment (tsnoam) [inline telegram/ext/filters.py]: that's a redundant empty line.

File: telegram/ext/filters.py [hunk 3]
```diff
@@ -171,6 +173,33 @@ def filter(self, message):
     command = _Command()
     """:obj:`Filter`: Messages starting with ``/``."""
 
+    class regex(BaseFilter):
```
Comment (tsnoam) [inline telegram/ext/filters.py]: Code for regex filter is good (regardless of some nitpicking).
However, it has nothing to do with the deep linking PR.
Please open a separate PR for it.

File: tests/test_helpers.py [hunk 4]
```diff
@@ -26,3 +27,30 @@ def test_escape_markdown(self):
...
+        expected = "https://t.me/{}?start={}".format(username, payload)
+        actual = helpers.create_deep_linked_url(username, payload)
+        assert expected == actual
+
+        payload = ""
+        expected = "https://t.me/{}".format(username)
+        assert expected == helpers.create_deep_linked_url(username, payload)
+        assert expected == helpers.create_deep_linked_url(username)
+
+        try:
```
Comment (tsnoam) [inline tests/test_helpers.py]: Please use `pytest.raises()` instead of the entire try...except... block

File: tests/test_helpers.py [hunk 5]
```diff
@@ -26,3 +27,30 @@ def test_escape_markdown(self):
...
+            # Invalid characters
+            helpers.create_deep_linked_url(username, 'text with spaces')
+            pytest.fail()
+        except Exception as e:
+            assert isinstance(e, ValueError)
+
+        try:
```
Comment (tsnoam) [inline tests/test_helpers.py]: Please use `pytest.raises()` instead of the entire try...except... block

File: telegram/ext/filters.py (hunk 1 above)
Comment (code1mountain) [reply telegram/ext/filters.py]: self.pattern can either be a regex compiled pattern or a string. You're solutions doesn't work for strings.

Comment (JosXa) [reply telegram/ext/filters.py]: This should be okay

File: telegram/ext/filters.py (hunk 3 above)
Comment (JosXa) [reply telegram/ext/filters.py]: See comment

**Comment by tsnoam** [comment]
@JosXa 
Please send Regex filter as a different PR and then we can proceed.

**Comment by JosXa** [comment]
#1028

**Comment by jh0ker** [comment]
@JosXa #1028 is merged

**Comment by JosXa** [comment]
@jh0ker Neat!

**Pull request closed without merging**